File: 1401-1500/1410-HTML Entity Parser/1410-HTML Entity Parser.cpp
class Solution {
public:
    string entityParser(string text) {
        int start = 0, curr = 0;
        unordered_map<string, char> table{{"&quot;", '\"'}, {"&apos;", '\''},{"&amp;", '&'},
                                          {"&gt;", '>'}, {"&lt;", '<'}, {"&frasl;", '/'}};
        for (char c : text) {
            text[curr] = c;
            if (c == '&') {
                start = curr;
            }
            else if (c == ';') {
                string s = text.substr(start, curr - start + 1);
                if (table.count(s)) {
                    text[start] = table[s];
                    curr = start;
                }
                start = curr + 1;
            }
            ++curr;
        }
        
        text.resize(curr);
        return text;
    }
};

File: 1401-1500/1410-HTML Entity Parser/1410-HTML Entity Parser.py
class Solution:
    def entityParser(self, text: str) -> str:
        start = curr = 0
        result = list(text)
        table = {"&quot;": '\"', "&apos;": '\'', "&amp;": '&',
                "&gt;": '>', "&lt;": '<', "&frasl;": '/'}
        
        for c in text:
            result[curr] = c
            if c == '&':
                start = curr
            elif c == ';' and 4 <= curr - start + 1 <= 7:
                s = ''.join(result[start: curr + 1])
                if s in table:
                    result[start] = table[s]
                    curr = start
                start = curr + 1
            curr += 1

        return ''.join(result[:curr])
